// npkill/lib/services/files/files.service.js
import fs, { accessSync, readFileSync, statSync } from 'fs';
import { readdir, stat } from 'fs/promises';
export class FileService {
    isValidRootFolder(path) {
        let stat;
        try {
            stat = statSync(path);
        }
        catch (error) {
            throw new Error('The path does not exist.');
        }
        if (!stat.isDirectory()) {
            throw new Error('The path must point to a directory.');
        }
        try {
            accessSync(path, fs.constants.R_OK);
        }
        catch (error) {
            throw new Error('Cannot read the specified path.');
        }
        return true;
    }
    convertKbToGB(kb) {
        const factorKBtoGB = 1048576;
        return kb / factorKBtoGB;
    }
    convertBytesToKB(bytes) {
        const factorBytestoKB = 1024;
        return bytes / factorBytestoKB;
    }
    convertGBToMB(gb) {
        const factorGBtoMB = 1024;
        return gb * factorGBtoMB;
    }
    getFileContent(path) {
        const encoding = 'utf8';
        return readFileSync(path, encoding);
    }
    isSafeToDelete(path, targetFolder) {
        return path.includes(targetFolder);
    }
    /** We consider a directory to be dangerous if it is hidden.
     *
     * > Why dangerous?
     * It is probable that if the node_module is included in some hidden directory, it is
     * required by some application like "spotify", "vscode" or "Discord" and deleting it
     * would imply breaking the application (until the dependencies are reinstalled).
     */
    isDangerous(path) {
        const hiddenFilePattern = /(^|\/)\.[^/.]/g;
        return hiddenFilePattern.test(path);
    }
    async getRecentModificationInDir(path) {
        const files = await this.getFileStatsInDir(path);
        const sorted = files.sort((a, b) => b.modificationTime - a.modificationTime);
        return sorted.length > 0 ? sorted[0].modificationTime : -1;
    }
    async getFileStatsInDir(dirname) {
        let files = [];
        const items = await readdir(dirname, { withFileTypes: true });
        for (const item of items) {
            if (item.isDirectory()) {
                if (item.name === 'node_modules') {
                    continue;
                }
                files = [
                    ...files,
                    ...(await this.getFileStatsInDir(`${dirname}/${item.name}`).catch(() => [])),
                ];
            }
            else {
                const path = `${dirname}/${item.name}`;
                const fileStat = await stat(path);
                files.push({ path, modificationTime: fileStat.mtimeMs / 1000 });
            }
        }
        return files;
    }
}
